src/table_mgr.cc

```diff
@@ -252,6 +246,57 @@ Status TableMgr::init(const TableMgrOptions& _options) {
             }
         }
 
+        // Adjust num L0 partitions blockingly only when it is not logSectionOnly mdoe,
+        // not read only mode, and the number of L0 partitions read from existing manifest
+        // file is different from the number specified in db config.
+        if (!db_config->logSectionOnly && !db_config->readOnly
+            && numL0Partitions != db_config->numL0Partitions) {
+            _log_info(myLog,
+                      "adjust numL0 partitions: %zu -> %zu",
+                      numL0Partitions,
+                      db_config->numL0Partitions);
+
+            if (!db_config->nextLevelExtension) {
+                _log_err(myLog, "[Adjust numL0] not allowed in L0 only mode");
+                throw Status(Status::INVALID_CONFIG);
+            }
+
+            // Need to compact all existing L0 tables to L1 and recreate empty L0 tables,
+            // otherwise hash will be messed up.
+            for (size_t ii = 0; ii < numL0Partitions; ++ii) {
+                // Force compact L0 table to L1 in blocking manner to reduce L0
+                // partitions.
+                std::list<TableInfo*> tables;
+                s = mani->getL0Tables(ii, tables);
+                if (tables.size() != 1 || !s) {
```

Comment:
There can be more than 1 tables for the same hash, if previous L0 table was in the middle of compaction and the DB was closed.